test/unit/offense_collection_test.rb

```diff
@@ -53,12 +57,22 @@ class OffenseCollectionTest < Minitest::Test
         .with(package, "./buyer/deprecated_references.yml")
         .returns(deprecated_references)
 
-      offense = Packwerk::ReferenceOffense.new(reference: reference, violation_type: ViolationType::Dependency)
+      offense = Packwerk::ReferenceOffense.new(
+        reference: reference,
+        violation_type: ViolationType::Dependency,
+        message: "some message"
+      )
+
       assert @offense_collection.listed?(offense)
     end
 
     test "#listed? returns false if constant is not listed in file " do
-      offense = Packwerk::ReferenceOffense.new(reference: build_reference, violation_type: ViolationType::Dependency)
+      offense = Packwerk::ReferenceOffense.new(
+        reference: build_reference,
+        violation_type: ViolationType::Dependency,
+        message: "some message"
```

Comment:
almost feels like there is a `Violation` type missing here that could generate the message on demand, so that instead of injecting a string, we would inject the capability to generate the string 🤔

Reply:
But that can be a separate PR, if you agree it makes sense. And maybe it is a premature abstraction, I haven't thought too much about it.

The smell for me is that we're injecting two things that are tightly coupled to each other. We would never pass a `ViolationType::Dependency` with a message for a privacy error.

Reply:
For sure, I think this speaks to Matt's thought: https://github.com/Shopify/packwerk/pull/196#discussion_r853115253

I think what might make most sense would be the checkers create the offense themselves rather than calling a bunch of separate checker interface methods. I do think there is value in having a reference offense have a `violation_type` as part of its API since I can see a client wanting to react differently to different types of violations.

I agree that we'd never want to pass a dependency violation type for a privacy error, although I'm not sure how we can prevent this systematically while simultaneously allowing clients to know what type of violation type it is 🤔